bin/varnishd/cache/cache_conn_pool.c

```diff
@@ -249,17 +254,32 @@ VCP_Rel(struct conn_pool **cpp)
 		(void)shutdown(pfd->fd, SHUT_RDWR);
 		cp->n_kill++;
 	}
-	while (cp->n_kill) {
-		Lck_Unlock(&cp->mtx);
-		(void)usleep(20000);
+
+	VRBT_INSERT(vrb, &dead_pools, cp);
```

Comment:
I just realized that this insertion may not happen in the CLI thread, so we need to guard it. Would the existing `conn_pools_mtx` be enough or do we need a new mutex for this tree?

This implies that we also need to guard the loop in `VCP_RelPoll()` with the same lock. The entire loop or just the VRBT operations?

Reply:
From what I've tested reusing `conn_pools_mtx` works fine
Since the loop is non-blocking I guess we can safely guard the whole loop

Reply:
I think we should lock before the loop, unlock after removing an entry, lock again at the end of the loop, and unlock after the loop.

Regarding `conn_pools_mtx` reuse one quick way to assess whether we can safely do that is running `make witness` after implementing the change. It creates a relationship between this lock and per-pool locks, if there isn't already one. You need to check that whenever those locks are taken, it's always in the same order.